### PR TITLE
fix: prevent `no_predicates` from removing predicates in calling function

### LINKED ISSUE #5435: #[no_predicates] generates correctness issues
### Aim

Using `#[no_predicates]` around poseidon since it shouldn't fail regardless of the input passed in.

### Expected Behavior

Functions wrapped in `#[no_predicates]` should not corrupt the program

### Bug

Here is a branch that switches poseidon to use `#[no_predicates]`. Tests that use poseidon start to fail:

https://github.com/AztecProtocol/aztec-packages/pull/7388
https://github.com/AztecProtocol/aztec-packages/actions/runs/9840072889/job/27164023595?pr=7388

Maybe @TomAFrench has a simpler reproduction since he diagnosed that this was the issue.

### To Reproduce

1.
2.
3.
4.

### Project Impact

None

### Impact Context

This and the switch to ultrahonk block poseidon2 usage

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Installation Method

None

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #5435 

## Summary\*

This PR caches the current `side_effects_enabled` value on the `PerFunctionContext` and restores it after inlining another function. This ensures that any `EnableSideEffects` instructions from the inlined function do not leak out into the calling function.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/inlining.rs [hunk 1]
```diff
@@ -482,12 +485,28 @@ impl<'function> PerFunctionContext<'function> {
...
+                                self.context.builder.insert_enable_side_effects_if(condition);
+                            }
                         } else {
                             self.push_instruction(*id);
                         }
                     }
                     None => self.push_instruction(*id),
                 },
+                Instruction::EnableSideEffects { condition } => {
+                    self.side_effects_enabled = Some(self.translate_value(*condition));
```
Comment (sirasistant) [inline compiler/noirc_evaluator/src/ssa/opt/inlining.rs:507]: I think we can just make it a function local variable

**Review by sirasistant — COMMENTED** [review]
Looks good, let me just pass the e2e tests in aztec-packages to double check

File: compiler/noirc_evaluator/src/ssa/opt/inlining.rs (hunk 1 above)
Comment (TomAFrench) [reply compiler/noirc_evaluator/src/ssa/opt/inlining.rs]: Yeah, we should be down a single block at the point we have these instructions having been inserted so this should be fine.

**Review by sirasistant — APPROVED** [review]
Perfect, it's passing e2e tests on aztec-packages too!

**Pull request merged**